yarn-project/end-to-end/src/e2e_ordering.test.ts

```diff
@@ -29,38 +29,52 @@ describe('e2e_ordering', () => {
   describe('with parent and child contract', () => {
     let parent: ParentContract;
     let child: ChildContract;
+    let pubSetValueSelector: Buffer;
 
     beforeEach(async () => {
       parent = await ParentContract.deploy(wallet).send().deployed();
       child = await ChildContract.deploy(wallet).send().deployed();
+      pubSetValueSelector = child.methods.pubSetValue.selector;
     });
 
-    const getChildStoredValue = () =>
-      aztecRpcServer.getPublicStorageAt(child.address, new Fr(1)).then(x => toBigInt(x!));
+    describe('enqueued public calls ordering', () => {
+      const nestedValue = 10n;
+      const directValue = 20n;
 
-    // Fails since current value at the end of execution is 10, not 20
-    it.skip('orders public function execution requests when nested call is last', async () => {
-      const actions: FunctionCall[] = [
-        child.methods.pubSetValue(10).request(),
-        parent.methods.enqueueCallToChild(child.address, child.methods.pubSetValue.selector, 20).request(),
-      ];
+      const expectedOrders = {
+        enqueueCallsToChildWithNestedFirst: [nestedValue, directValue],
+        enqueueCallsToChildWithNestedLast: [directValue, nestedValue],
+      } as const;
 
-      await new BatchCall(wallet, actions).send().wait();
-      expect(await getChildStoredValue()).toEqual(20n);
-      const logs = await aztecRpcServer.getUnencryptedLogs(1, 10).then(L2BlockL2Logs.unrollLogs);
-      expect(logs).toEqual([[10], [20]].map(Buffer.from));
-    });
+      it.each(['enqueueCallsToChildWithNestedFirst', 'enqueueCallsToChildWithNestedLast'] as const)(
+        'orders public function execution in %s',
+        async method => {
+          const expectedOrder = expectedOrders[method];
+          const action = parent.methods[method](child.address, pubSetValueSelector);
+          const tx = await action.simulate();
+          await action.send().wait();
+
+          // There are two enqueued calls
+          const enqueuedPublicCalls = tx.enqueuedPublicFunctionCalls;
+          expect(enqueuedPublicCalls.length).toEqual(2);
+
+          // The call stack hashes in the output of the kernel proof match the tx enqueuedPublicFunctionCalls
+          const hashes = await Promise.all(enqueuedPublicCalls.map(c => c.toPublicCallStackItem().then(i => i.hash())));
+          expect(tx.data.end.publicCallStack.slice(0, 2)).toEqual(hashes);
+
+          // The enqueued public calls are in the expected order based on the argument they set (stack is reversed!)
+          expect(enqueuedPublicCalls.map(c => c.args[0].toBigInt())).toEqual([...expectedOrder].reverse());
```

Comment:
Why are they reversed? Which one has them backwards?

Reply:
Kernel pops the next thing to work on from the end of the array, so the list of calls is in the reverse order you'd expect.

Reply:
Ahh thank you for the clarification!